<launch>
    <arg name="robot_name" default="robot_0" />
    <arg name="init_x" default="10.0" />
    <arg name="init_y" default="1.0" />
    <arg name="init_a" default="1.57" />
    <arg name="init_cov" default="0.25" />
    <arg name="robotPort"/>
    <arg name="laserPort"/>
    <arg name="laserPort2"/>
    <arg name="joyPort" default="/dev/input/js0"/>
    <arg name="use_move_base" default="true" />
    <arg name="use_amcl" default="true" />
    
    <param name="emergency_stop" type="int" value="0" />
    
    <group ns="$(arg robot_name)">
        <param name="robot_name" value="$(arg robot_name)" />
        <param name="tf_prefix" value="$(arg robot_name)" />

        <!-- AMCL -->
    
        <group if="$(arg use_amcl)">
    
            <node pkg="amcl" type="amcl" name="amcl" respawn="true">
                
                <remap from="static_map" to="/static_map" />
                
                <!-- Default parameters -->
                <rosparam file="$(find stage_environments)/config/sapienzbot/localization/amcl.yaml" command="load" />
                
                <!-- Override AMCL Frame Params to include prefix -->
                <param name="base_frame_id" value="/$(arg robot_name)/base_link"/>
                <param name="odom_frame_id" value="/$(arg robot_name)/odom"/>
                <param name="global_frame_id" value="/map"/> <!--common map frame for all robots -->

                <param name="initial_pose_x" value="$(arg init_x)" />
                <param name="initial_pose_y" value="$(arg init_y)" />
                <param name="initial_pose_a" value="$(arg init_a)" />
                <param name="initial_cov_xx" value="0.25" />
                <param name="initial_cov_yy" value="0.25" />
                <param name="initial_cov_aa" value="0.25" />
                
            </node>
        
        </group>
        
        
        
        <!-- MOVE BASE -->
        
        <group if="$(arg use_move_base)">
            <node pkg="move_base" type="move_base" respawn="false" name="move_base_node" output="screen">
                <remap from="map" to="/map" />
                <remap from="odom" to="odom" />
                <remap from="cmd_vel" to ="desired_cmd_vel" />
                
                <rosparam file="$(find stage_environments)/config/sapienzbot/navigation/move_base_params.yaml" command="load" />
                <rosparam file="$(find stage_environments)/config/sapienzbot/navigation/costmap_common_params.yaml" command="load" ns="global_costmap" />
                <rosparam file="$(find stage_environments)/config/sapienzbot/navigation/costmap_common_params.yaml" command="load" ns="local_costmap" />
                <rosparam file="$(find stage_environments)/config/sapienzbot/navigation/local_costmap_params.yaml" command="load" />
                <rosparam file="$(find stage_environments)/config/sapienzbot/navigation/global_costmap_params.yaml" command="load" />
                <rosparam file="$(find stage_environments)/config/sapienzbot/navigation/base_local_planner_params.yaml" command="load" />
                
                <!-- Override MOVE_BASE Frame Params to include prefix -->
                <param name="global_costmap/laser_scan_sensor/sensor_frame" value="/$(arg robot_name)/base_laser_link"/>
                <param name="global_costmap/laser_scan_sensor/topic" value="/$(arg robot_name)/scan"/>
                <param name="global_costmap/robot_base_frame" value="/$(arg robot_name)/base_link"/>   
                <param name="local_costmap/global_frame" value="/$(arg robot_name)/odom"/>
                <param name="local_costmap/laser_scan_sensor/sensor_frame" value="/$(arg robot_name)/base_laser_link"/>
                <param name="local_costmap/laser_scan_sensor/topic" value="/$(arg robot_name)/scan"/>
                <param name="local_costmap/robot_base_frame" value="/$(arg robot_name)/base_link"/>
                
            </node>
        </group>
    
    
        <node pkg="living_at_diag" name="getpose" type="getpose.py" args="-srv" output="screen"/>
        <node pkg="emergency_stop" name="emerg" type="emergency_stop" args="" output="screen"/>

        <node pkg="laser_analysis" type="laserobstacle" args=""  name="laserobstacle" output="screen">
            <param name="~laser_frame_id" value="$(arg robot_name)/base_laser_link" />
        </node>

        <node name="joy_node" pkg="joy" type="joy_node" output="screen" >
            <param name="~deadzone" type="double" value="0.2"/> 
            <param name="~autorepeat_rate" type="double" value="20."/> 
            <param name="~dev" type="str" value="/dev/input/js0"/>      
        </node>

        <node name="joy2twist" pkg="human_interface" type="joy2twist.py" output="screen" >
            <remap from="cmd_vel" to ="desired_cmd_vel" />
        </node>

	<node name="link_broadcaster_2" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 /$(arg robot_name)/base_link /$(arg robot_name)/base_frame 0.2" />
        <node name="link_broadcaster_3" pkg="tf" type="static_transform_publisher" args="-0.2 0 0.24 0 -0.698 3.14 /$(arg robot_name)/base_link /$(arg robot_name)/top_link 1" />
        <node name="link_broadcaster_4" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 /$(arg robot_name)/top_link /camera_link 1" />        
      </group>

</launch>

